packages/pwa-buildpack/lib/WebpackTools/plugins/RootComponentsPlugin.js

```diff
@@ -0,0 +1,217 @@
+        }).apply(this.compiler);
+        new ProvidePlugin({
+            fetchRootComponent: 'FETCH_ROOT_COMPONENT'
+        }).apply(this.compiler);
+    }
+    findRootComponentsIn(dir) {
+        return new Promise(resolve => {
+            const jsFiles = [];
+            const done = () => resolve(jsFiles);
+            const fs = this.fs;
+            walk(dir, {
+                fs: {
+                    readdir: fs.readdir.bind(fs),
+                    stat: fs.stat.bind(fs)
+                }
+            })
+                .on('readable', function() {
+                    let item;
+                    while ((item = this.read())) {
+                        if (
+                            item.stats.isFile() &&
+                            extensionRE.test(item.path)
+                        ) {
+                            debug(`will check ${item.path} for RootComponent`);
+                            jsFiles.push(item.path);
+                        }
+                    }
+                })
+                .on('error', done)
+                .on('end', done);
+        });
+    }
+    async buildFetchModule() {
+        const { context, rootComponentsDirs } = this.opts;
+
+        // Create a list of absolute paths for root components. When a
+        // relative path is found, resolve it from the root context of
+        // the webpack build
+        const rootComponentsDirsAbs = rootComponentsDirs.map(dir =>
+            isAbsolute(dir) ? dir : join(context, dir)
+        );
+        debug(
+            'buildFetchModule: absolute root component dirs: %o',
+            rootComponentsDirsAbs
+        );
+        const rootComponentImporters = await rootComponentsDirsAbs.reduce(
+            async (importersPromise, rootComponentDir) => {
+                debug('gathering files from %s', rootComponentDir);
+                const [importerSources, rootComponentFiles] = await Promise.all(
+                    [
+                        importersPromise,
+                        this.findRootComponentsIn(rootComponentDir)
+                    ]
+                );
+                debug(
+                    'files from %s: %o',
+                    rootComponentDir,
+                    rootComponentFiles
+                );
+                await Promise.all(
+                    rootComponentFiles.map(async rootComponentFile => {
+                        debug('reading file %s', rootComponentFile);
+                        const fileContents = await this.readFile(
+                            rootComponentFile
+                        );
+                        const rootComponentSource =
+                            typeof fileContents !== 'string'
+                                ? fileContents
+                                : fileContents.toString('utf8');
+                        debug(
+                            'parsing %s source for directives',
+                            rootComponentSource.slice(0, 80)
```

Comment:
Why not just log the filename? Current output:

```
  pwa-buildpack:WebpackTools:plugins:RootComponentsPlugin.js parsing /**
 * @RootComponent
 * description = 'Basic CMS Page'
 * pageTypes = CMS_PAGE
 source for directives +2ms
```

Reply:
Oh, this is an actual directive `@RootComponent`.

Reply:
Yeah, we want to see the actual directive.